trieste/acquisition/rule.py

```diff
@@ -1889,6 +1896,304 @@ def get_dataset_min(
+        :param region_index: The index of the region in a multi-region search space. This is used to
+            identify the local models and datasets to use for acquisition. If `None`, the
+            global models and datasets are used.
+        """
+        assert len(regions) > 0, "at least one region should be provided"
+
+        # If set, assert all regions have the same index and matching the product index.
+        if region_index is not None:
+            assert all(
+                region.region_index == region_index
+                for region in regions
+                if region.region_index is not None
+            ), (
+                "all regions should have the same index, if set, as the "
+                f"product region ({region_index})"
+            )
+        else:
+            assert all(region.region_index is None for region in regions), (
+                f"regions can only have a region_index if the product region ({region_index}) "
+                "has one"
+            )
+
+        self._global_search_space = TaggedProductSearchSpace(
+            [region.global_search_space for region in regions]
+        )
+
+        # Find the most general dtype of all the regions.
+        dtypes = [np.array(region.location).dtype for region in regions]
+        most_general_dtype = np.result_type(*dtypes)
+        self.dtype = most_general_dtype
+
+        TaggedProductSearchSpace.__init__(self, regions)
+        # When UpdatableTrustRegion sets the region_index, it will also set the region_index for
+        # each region.
+        UpdatableTrustRegion.__init__(self, region_index)
+
+    @property
+    def region_index(self) -> Optional[int]:
+        """The index of the region in a multi-region search space."""
+        return self._region_index
+
+    @region_index.setter
+    def region_index(self, region_index: Optional[int]) -> None:
+        """Set the index of the region in a multi-region search space, including all sub-regions."""
+        self._region_index = region_index
+        # Override the region index for each sub-region. These would either already be set to the
+        # same value (assert in __init__), or None.
+        for region in self.regions.values():
+            region.region_index = region_index
+
+    @property
+    def regions(self) -> Mapping[str, UpdatableTrustRegionWithGlobalSearchSpace]:
+        """The sub-regions of the product trust region."""
+        _regions = {}
+        for tag, region in self._spaces.items():
+            assert isinstance(region, (UpdatableTrustRegionBox, UpdatableTrustRegionDiscrete))
+            _regions[tag] = region
+        return _regions
+
+    @property
+    def location(self) -> TensorType:
+        """
+        The location of the product trust region, concatenated from the locations of the
+        sub-regions.
+        """
+        return tf.concat(
+            [tf.cast(region.location, self.dtype) for region in self.regions.values()], axis=-1
```

Comment:
Not necessarily floats, whatever is the least general common type, but yes likely floats in most cases.

We already have this requirement in many places in existing code; for instance, wherever we use `TaggedProductSearchSpace`. For example, when we call `sample` in the optimiser -- it combines the samples by calling `tf.concat`. I think we force the user to create all search spaces with a common type, i.e. float64.

Here I am just being extra nice by doing the casting for the user if they didn't create the spaces with the same type. But I agree this is of limited value, since other places in our code will fallover anyway. So I can remove this for consistency sake with other code :-)

Reply:
no need to remove it then :)

but note about putting a warning in the docs is still valid

Reply:
It's fine, I have removed it. It convinced me that the added complexity was not worth it.

I have added a note about dtypes in 3 obvious places, `TaggedProductSearchSpace`, `UpdatableTrustRegionProduct` and `mixed_search_spaces.pct.py`. I wasn't sure about adding it to `Dataset`, since that doesn't directly reference mixed search spaces.